[#11571] Selective Deadline Extensions - DeadlineExtension entity and sending of emails

samuelfangjw opened the pull request:
Part of #11571 

**Outline of Solution**
This PR includes
- Add new DeadlineExtension entity and related methods and classes
- Task queue worker to create/update/delete deadline extensions and send appropriate emails
- Modifying automated closing reminder emails to send emails to users with deadline extensions
- Create and Delete Deadline Extension API for use in future e2e tests

jayasting98 reviewed (commented):
Largely good work. I couldn't find much problems with the code right now, just maybe a few things.

samuelfangjw commented on src/test/java/teammates/storage/api/DeadlineExtensionsDbTest.java:
```diff
@@ -0,0 +1,405 @@
+                .build();
+        InvalidParametersException ipe = assertThrows(InvalidParametersException.class,
+                () -> deadlineExtensionsDb.updateDeadlineExtension(invalidEmailUpdateOptions));
+        System.out.println(ipe.getMessage());
```
Yes, should be removed. Thanks for spotting this.

samuelfangjw commented on src/main/java/teammates/logic/api/EmailGenerator.java:
```diff
@@ -466,6 +483,49 @@ private EmailWrapper generateSessionLinksRecoveryEmailForExistingStudent(String
+        List<DeadlineExtensionAttributes> instructorDeadlines =
+                deadlineExtensions.stream().filter(x -> x.getIsInstructor()).collect(Collectors.toList());
+
+        boolean isEmailNeededForStudents =
+                !studentDeadlines.isEmpty() && fsLogic.isFeedbackSessionForUserTypeToAnswer(session, false);
+        boolean isEmailNeededForInstructors =
+                !instructorDeadlines.isEmpty() && fsLogic.isFeedbackSessionForUserTypeToAnswer(session, true);
+
+        List<StudentAttributes> students = new ArrayList<>();
+        if (isEmailNeededForStudents) {
+            for (DeadlineExtensionAttributes studentDeadline : studentDeadlines) {
```
I'm optimizing to minimize db reads here. If I'm not wrong, `isFeedbackSessionForUserTypeToAnswer` requires fetching multiple `FeedbackQuestion` entities from db.

jayasting98 replied:
Ahhh ok. That makes sense. Then it's all good then.

samuelfangjw commented on src/main/java/teammates/logic/api/EmailGenerator.java:
```diff
@@ -587,6 +587,65 @@ private EmailWrapper generateSessionLinksRecoveryEmailForExistingStudent(String
+        return generateDeadlineExtensionEmail(course, feedbackSession, null, instructor, endTime, emailType);
+    }
+
+    private EmailWrapper generateDeadlineExtensionEmail(
+            CourseAttributes course, FeedbackSessionAttributes session, StudentAttributes student,
+            InstructorAttributes instructor, Instant endTime, EmailType emailType) {
```
Might be possible to refactor to fetch student/instructor in this method if that's neater. I don't think student/instructor is required anywhere else.

samuelfangjw replied:
I've refactored it to get rid of issue, please have a look when you have time :)

jayasting98 replied:
Yea I think that works.

moziliar commented on src/test/java/teammates/ui/webapi/DeadlineExtensionsWorkerActionTest.java:
```diff
@@ -0,0 +1,181 @@
+
+        List<TaskWrapper> tasksAdded = mockTaskQueuer.getTasksAdded();
+        String datetimeDisplayFormat = "EEE, dd MMM yyyy, hh:mm a z";
+        for (TaskWrapper task : tasksAdded) {
```
`var task`?

moziliar commented on src/test/java/teammates/ui/webapi/DeadlineExtensionsWorkerActionTest.java:
```diff
@@ -0,0 +1,181 @@
+
+        ______TS("notifyUsers set to false; no tasks added");
+
+        logic.deleteDeadlineExtension(courseId, feedbackSessionName, studentCreate.getEmail(), false);
+        logic.deleteDeadlineExtension(courseId, feedbackSessionName, instructorCreate.getEmail(), true);
+        removeAndRestoreTypicalDataBundle();
```
Why delete extension before `removeAndRestore`?

Also, after these are done, deadlines for `studentCreate` and `instructorCreate` are both the `newEndTime`?

samuelfangjw replied:
>Why delete extension before removeAndRestore?

My bad, made some wrong assumptions here. 

> Also, after these are done, deadlines for studentCreate and instructorCreate are both the newEndTime?

`studentCreate` and `instructorCreate` weren't modified. I don't believe there will be any changes.

moziliar commented on src/test/java/teammates/ui/webapi/DeleteDeadlineExtensionActionTest.java:
```diff
@@ -0,0 +1,69 @@
+    @Override
+    @Test
+    protected void testExecute() {
+        DeadlineExtensionAttributes deadlineExtension = typicalBundle.deadlineExtensions.get("student4InCourse1Session1");
```
Has this been added to the "db" and is ensured that it exists prior to delete action execution?

samuelfangjw replied:
Will add a check to verify existence of deadline of deadline extension for good measure.

moziliar commented on src/test/java/teammates/ui/webapi/DeleteDeadlineExtensionActionTest.java:
```diff
@@ -0,0 +1,69 @@
+
+        ______TS("Not enough parameters");
+
+        verifyHttpParameterFailure();
```
Try all possible comp?

samuelfangjw replied:
Added 4 testcases, one for each missing param. All possible combinations will be a bit too much.

moziliar commented on src/test/java/teammates/ui/webapi/FeedbackSessionClosingRemindersActionTest.java:
```diff
@@ -103,14 +105,50 @@ public void testExecute() throws Exception {
         verifyPresentInDatabase(session3);
 
+        // update deadline extensions to have end time within the next 24 hours
+        DeadlineExtensionAttributes deadlineExtensionStudent =
+                typicalBundle.deadlineExtensions.get("student4InCourse1Session1");
+        logic.updateDeadlineExtension(
+                DeadlineExtensionAttributes
+                        .updateOptionsBuilder(
+                            deadlineExtensionStudent.getCourseId(),
+                            deadlineExtensionStudent.getFeedbackSessionName(),
+                            deadlineExtensionStudent.getUserEmail(),
+                            deadlineExtensionStudent.getIsInstructor())
+                        .withEndTime(TimeHelper.getInstantHoursOffsetFromNow(12))
+                        .build());
+        DeadlineExtensionAttributes deadlineExtensionStudentAlreadyAnswered =
+                typicalBundle.deadlineExtensions.get("student3InCourse1Session1");
+        logic.updateDeadlineExtension(
+                DeadlineExtensionAttributes
+                        .updateOptionsBuilder(
+                            deadlineExtensionStudentAlreadyAnswered.getCourseId(),
+                            deadlineExtensionStudentAlreadyAnswered.getFeedbackSessionName(),
+                            deadlineExtensionStudentAlreadyAnswered.getUserEmail(),
+                            deadlineExtensionStudentAlreadyAnswered.getIsInstructor())
+                        .withEndTime(TimeHelper.getInstantHoursOffsetFromNow(12))
+                        .build());
+        DeadlineExtensionAttributes deadlineExtensionInstructor =
+                typicalBundle.deadlineExtensions.get("instructor2InCourse1Session1");
+        logic.updateDeadlineExtension(
+                DeadlineExtensionAttributes
+                        .updateOptionsBuilder(
+                            deadlineExtensionInstructor.getCourseId(),
+                            deadlineExtensionInstructor.getFeedbackSessionName(),
+                            deadlineExtensionInstructor.getUserEmail(),
+                            deadlineExtensionInstructor.getIsInstructor())
+                        .withEndTime(TimeHelper.getInstantHoursOffsetFromNow(12))
+                        .build());
```
A bit verbose; can you try to make a list of the users and then use a for loop to update their `endTime`?

```java
DeadlineExtensionAttributes deadlineExtensionStudent =
        typicalBundle.deadlineExtensions.get("student4InCourse1Session1");       
DeadlineExtensionAttributes deadlineExtensionStudentAlreadyAnswered =
        typicalBundle.deadlineExtensions.get("student3InCourse1Session1");
DeadlineExtensionAttributes deadlineExtensionInstructor =
        typicalBundle.deadlineExtensions.get("instructor2InCourse1Session1");
List<DeadlineExtensionAttributes> deadlinesToExtend = List.of(
        deadlineExtensionStudent, 
        deadlineExtensionStudentAlreadyAnswered,
        deadlineExtensionInstructor);
for (var deadlineExtension : deadlinesToExtend) {
    ...
}
```

moziliar commented on src/test/java/teammates/ui/webapi/GetDeadlineExtensionActionTest.java:
```diff
@@ -0,0 +1,79 @@
+
+        ______TS("Not enough parameters");
+
+        verifyHttpParameterFailure();
```
ditto

moziliar commented on src/test/java/teammates/ui/webapi/GetDeadlineExtensionActionTest.java:
```diff
@@ -0,0 +1,79 @@
+
+        DeadlineExtensionData response = (DeadlineExtensionData) r.getOutput();
+
+        assertEquals(deadlineExtension.getCourseId(), response.getCourseId());
+        assertEquals(deadlineExtension.getFeedbackSessionName(), response.getFeedbackSessionName());
+        assertEquals(deadlineExtension.getUserEmail(), response.getUserEmail());
+        assertEquals(deadlineExtension.getIsInstructor(), response.getIsInstructor());
```
For ID fields we can use `.equals`?

samuelfangjw replied:
Are you suggesting using `assertTrue(x.equals(y))`?

moziliar replied:
At least for identifier fields, we can do so? Not sure if this makes the code less readable.

samuelfangjw replied:
This is possible for the first 3 (except `isInstructor` which is a primitive boolean value). 

Implementation wise, it doesn't make much of a difference, except that `assertEquals` is null safe although that doesn't come into play here. Both use `.equals` to compare objects.

Comes down to readability I think. Personally prefer `assertEquals` here because it
1. Gives better error message when condition fails
2. Makes it explicit that we are checking for equality

What's the rationale behind using `assertTrue(x.equals(y))` instead?

moziliar replied:
I see. I do feel that the readability of the test failure will be worse.

The rationale is to compare id fields then other fields, as id fields should be the same regardless?

samuelfangjw replied:
Sorry I don't see the difference here. This tests that "id fields" are the same as well? There shouldn't be any difference between `assertEquals` and `.equals` for non-null fields, since they both use `.equals` to compare. Are you referring to comparing identity using `==` instead?

moziliar replied:
I think nvm... The current way is more straightforward.

moziliar reviewed (changes requested):
Great job! The code is really neat!

I left some nit comments. Impl LGTM.

samuelfangjw commented on src/main/java/teammates/storage/api/DeadlineExtensionsDb.java:
```diff
@@ -0,0 +1,209 @@
+        }
+
+        if (updateOptions.getUserEmail().equals(deadlineExtension.getUserEmail())) {
+            saveEntity(deadlineExtension.toEntity());
```
I believe the check will be redundant. I don't believe there is any case where the update method is called but the update is not needed.

That being said, I don't think it hurts to add either.

samuelfangjw commented on src/main/java/teammates/common/datatransfer/attributes/DeadlineExtensionAttributes.java:
```diff
@@ -0,0 +1,312 @@
+                    && Objects.equals(this.feedbackSessionName, otherDeadlineExtension.feedbackSessionName)
+                    && Objects.equals(this.userEmail, otherDeadlineExtension.userEmail)
+                    && Objects.equals(this.isInstructor, otherDeadlineExtension.isInstructor);
+        } else {
+            return false;
+        }
```
Yes I believe so, for cases where the objects are not of the same type.

samuelfangjw commented on src/test/java/teammates/common/datatransfer/attributes/DeadlineExtensionAttributesTest.java:
```diff
@@ -0,0 +1,214 @@
+
+    @Test
+    public void testHashCode() {
+        DeadlineExtensionAttributes deadlineExtension = getValidDeadlineExtensionAttributesObject(true);
+
+        // When the two deadline extensions have same values, they should have the same hash code
+        DeadlineExtensionAttributes deadlineExtensionSimilar = getValidDeadlineExtensionAttributesObject(true);
+
+        assertTrue(deadlineExtension.hashCode() == deadlineExtensionSimilar.hashCode());
+
+        // When the two deadline extensions are different, they should have different hash code
+        DeadlineExtensionAttributes deadlineExtensionDifferent = getValidDeadlineExtensionAttributesObject(false);
+
+        assertFalse(deadlineExtension.hashCode() == deadlineExtensionDifferent.hashCode());
```
Added one more test case

samuelfangjw commented on src/main/java/teammates/storage/api/DeadlineExtensionsDb.java:
```diff
@@ -0,0 +1,209 @@
+        assert oldEmail != null;
+        assert newEmail != null;
+
+        List<DeadlineExtension> entitiesToUpdate = load().project()
+                .filter("courseId =", courseId)
+                .filter("userEmail =", oldEmail)
+                .filter("isInstructor =", isInstructor)
+                .list();
+
+        for (DeadlineExtension deadlineExtension : entitiesToUpdate) {
+            try {
+                updateDeadlineExtension(DeadlineExtensionAttributes
+                        .updateOptionsBuilder(courseId, deadlineExtension.getFeedbackSessionName(), oldEmail, isInstructor)
+                        .withNewEmail(newEmail)
+                        .build());
+            } catch (EntityDoesNotExistException ednee) {
+                // ignore: deadline extension entity already deleted
+            }
+        }
+    }
```
This will make things more messy, have to account for `InvalidParametersException` in the anonymous function.

samuelfangjw commented on src/main/java/teammates/storage/api/DeadlineExtensionsDb.java:
```diff
@@ -0,0 +1,209 @@
+        if (query.isFeedbackSessionNamePresent()) {
+            entitiesToDelete = entitiesToDelete.filter("feedbackSessionName =", query.getFeedbackSessionName());
+        }
+
```
`isInstructor` is not needed for this particular query.

samuelfangjw commented on src/main/java/teammates/storage/api/DeadlineExtensionsDb.java:
```diff
@@ -0,0 +1,209 @@
+        deleteEntity(entitiesToDelete.keys().list());
+    }
+
+    /**
+     * Deletes all deadline extensions of a user in a course.
+     */
+    public void deleteDeadlineExtensions(String courseId, String userEmail, boolean isInstructor) {
+        assert courseId != null;
+        assert userEmail != null;
+
+        Query<DeadlineExtension> entitiesToDelete = load().project();
+        entitiesToDelete = entitiesToDelete.filter("courseId =", courseId);
+        entitiesToDelete = entitiesToDelete.filter("userEmail =", userEmail);
+        entitiesToDelete = entitiesToDelete.filter("isInstructor =", isInstructor);
+
+        deleteEntity(entitiesToDelete.keys().list());
+    }
```
I combined them together. Had to make changes to `AttributeDeletionQuery` too.  No change to index, only 3 possible combinations of attributes.

1. Course ID only
2. Course ID and FeedbackSessionName
3. Course ID and UserEmail and IsInstructor

moziliar replied:
Looks better now!

samuelfangjw commented:
I've updated the emails for deadline extensions to show the previous deadline as well as the new deadline.

samuelfangjw commented:
Hi @damithc, could you kindly help to check if the following emails for deadline extensions are ok?

**When a user is given a deadline:**
Subject: TEAMMATES: Deadline extension given [Course: %s][Feedback Session: %s]
<img width="743" alt="Screen Shot 2022-03-14 at 1 17 26 PM" src="https://user-images.githubusercontent.com/60355570/158109367-005cfff1-8129-4e0c-befc-8272b4682428.png">

**When a user's deadline is updated:**
Subject: TEAMMATES: Deadline extension updated [Course: %s][Feedback Session: %s]
<img width="743" alt="Screen Shot 2022-03-14 at 1 17 18 PM" src="https://user-images.githubusercontent.com/60355570/158109394-389dcca4-077f-46dc-b154-b8cf93ce6f43.png">

**When a user's deadline is revoked:**
Subject: TEAMMATES: Deadline extension revoked [Course: %s][Feedback Session: %s]
<img width="743" alt="Screen Shot 2022-03-14 at 1 17 21 PM" src="https://user-images.githubusercontent.com/60355570/158109420-07d7ac1b-550a-4559-9a50-f3c56f6f278c.png">

Could I also check if the language used is suitable (given/updated/revoked).

Thank you!

damithc commented:
> Could I also check if the language used is suitable (given/updated/revoked).

I'm fine with these, @samuelfangjw

moziliar commented on src/test/java/teammates/logic/core/DeadlineExtensionsLogicTest.java:
```diff
@@ -0,0 +1,418 @@
+                deadlineExtensionTwelveHoursAhead, deadlineExtensionOneDayAhead,
+                deadlineExtensionInstructor, deadlineExtensionOneDayBefore, deadlineExtensionEmailSent);
+
+        for (DeadlineExtensionAttributes deadlineExtension : deadlineExtensions) {
```
`var deadlineExtension`

moziliar commented on src/test/java/teammates/logic/core/DeadlineExtensionsLogicTest.java:
```diff
@@ -0,0 +1,418 @@
+        assertFalse(deadlineExtensionsNeedingClosing.contains(deadlineExtensionOneDayBefore));
+        assertFalse(deadlineExtensionsNeedingClosing.contains(deadlineExtensionEmailSent));
+
+        for (DeadlineExtensionAttributes deadlineExtension : deadlineExtensionsNeedingClosing) {
```
Ditto

moziliar commented on src/test/java/teammates/storage/api/DeadlineExtensionsDbTest.java:
```diff
@@ -0,0 +1,432 @@
+        ______TS("typical success case: delete all deadline extensions in course");
+
+        query = AttributesDeletionQuery.builder()
+                        .withCourseId(validCourseId)
```
Indent?

moziliar commented on src/test/java/teammates/ui/webapi/DeleteDeadlineExtensionActionTest.java:
```diff
@@ -0,0 +1,103 @@
+
+        ______TS("Missing parameter");
+
+        String[] params = new String[] {
+                Const.ParamsNames.FEEDBACK_SESSION_NAME, deadlineExtension.getFeedbackSessionName(),
+                Const.ParamsNames.USER_EMAIL, deadlineExtension.getUserEmail(),
+                Const.ParamsNames.IS_INSTRUCTOR, Boolean.toString(deadlineExtension.getIsInstructor()),
+        };
+
+        verifyHttpParameterFailure(params);
+
+        params = new String[] {
+                Const.ParamsNames.COURSE_ID, deadlineExtension.getCourseId(),
+                Const.ParamsNames.USER_EMAIL, deadlineExtension.getUserEmail(),
+                Const.ParamsNames.IS_INSTRUCTOR, Boolean.toString(deadlineExtension.getIsInstructor()),
+        };
+
+        verifyHttpParameterFailure(params);
+
+        params = new String[] {
+                Const.ParamsNames.COURSE_ID, deadlineExtension.getCourseId(),
+                Const.ParamsNames.FEEDBACK_SESSION_NAME, deadlineExtension.getFeedbackSessionName(),
+                Const.ParamsNames.IS_INSTRUCTOR, Boolean.toString(deadlineExtension.getIsInstructor()),
+        };
+
+        verifyHttpParameterFailure(params);
+
+        params = new String[] {
+                Const.ParamsNames.COURSE_ID, deadlineExtension.getCourseId(),
+                Const.ParamsNames.FEEDBACK_SESSION_NAME, deadlineExtension.getFeedbackSessionName(),
+                Const.ParamsNames.USER_EMAIL, deadlineExtension.getUserEmail(),
+        };
```
Would it be good to use commented lines to indicate which parameter is missing?
`// Const.ParamsNames.USER_EMAIL`

moziliar commented on src/test/java/teammates/ui/webapi/FeedbackSessionClosingRemindersActionTest.java:
```diff
@@ -103,14 +105,40 @@ public void testExecute() throws Exception {
+                deadlineExtensionStudentAlreadyAnswered,
+                deadlineExtensionInstructor);
+
+        for (DeadlineExtensionAttributes deadlineExtension : deadlineExtensions) {
```
var or forEach?

samuelfangjw replied:
We should be careful not to overuse var everywhere. There isn't really much value to replacing explicit types with var in such cases, especially if the line is fairly short. In addition, var is not used consistently throughout the project. In this case, explicit types still provide extra contextual information. Is this `DeadlineExtension` or `DeadlineExtensionAttributes`?

moziliar replied:
Im only raising this because the type is literally at line 115. This kind is ok to me, and we should start using Java 11 features?

moziliar replied:
It is communicated with Wilson that this can be used discreetly. I do believe having a type here when it is clear that `deadlineExtensions` is a `List<DeadlineExtensions>` is redundant.

moziliar commented on src/test/java/teammates/ui/webapi/GetDeadlineExtensionActionTest.java:
```diff
@@ -0,0 +1,113 @@
+                Const.ParamsNames.COURSE_ID, deadlineExtension.getCourseId(),
+                Const.ParamsNames.FEEDBACK_SESSION_NAME, deadlineExtension.getFeedbackSessionName(),
+                Const.ParamsNames.USER_EMAIL, deadlineExtension.getUserEmail(),
+        };
```
Ditto

moziliar commented on src/main/java/teammates/common/datatransfer/attributes/DeadlineExtensionAttributes.java:
```diff
@@ -0,0 +1,312 @@
+        DeadlineExtension deadlineExtension = new DeadlineExtension(getCourseId(), getFeedbackSessionName(),
+                getUserEmail(), getIsInstructor(), getSentClosingEmail(), getEndTime());
+
+        if (!Const.TIME_REPRESENTS_DEFAULT_TIMESTAMP.equals(this.getCreatedAt())) {
```
Thanks! Both sounds good. Can leave `this` be as it reads better. Just trying to understand.

samuelfangjw commented on src/main/java/teammates/storage/api/DeadlineExtensionsDb.java:
```diff
@@ -0,0 +1,208 @@
+        assert courseId != null;
+        assert oldEmail != null;
+        assert newEmail != null;
+
```
Shouldn't be an issue, will update.

samuelfangjw commented on src/main/java/teammates/storage/api/DeadlineExtensionsDb.java:
```diff
@@ -0,0 +1,208 @@
+
+        Query<DeadlineExtension> entitiesToDelete = load().project().filter("courseId =", query.getCourseId());
+
+        if (query.isFeedbackSessionNamePresent()) {
+            entitiesToDelete = entitiesToDelete.filter("feedbackSessionName =", query.getFeedbackSessionName());
+        } else if (query.isUserEmailPresent() && query.isIsInstructorPresent()) {
+            entitiesToDelete = entitiesToDelete.filter("userEmail =", query.getUserEmail());
+            entitiesToDelete = entitiesToDelete.filter("isInstructor =", query.getIsInstructor());
+        }
```
Only three possible cases, 
1. courseId
2. courseId, fsName
3. courseId, userEmail, isInstructor

samuelfangjw replied:
In this case, I believe they will have to modify the index to add this particular combination of attributes anyways?

moziliar replied:
In some roundabout case, a contributor who's unaware of the implicit message here might attempt to send in incorrect combination like only `isInstructor` is present.
The result might be that nothing is filtered and all extensions within the course are gone. 

Would expect some sanity check somewhere?

samuelfangjw replied:
`might attempt to send in incorrect combination like only isInstructor is present`

I believe the assertion above this should make it obvious that this is not possible. I could add another assertion for valid combination if this is a concern.

samuelfangjw replied:
You mentioned only `isInstructor` in your previous comment.

Nevertheless, I take your point. Will add a validity check for this.

moziliar replied:
Ah, sorry about my phrasing. Some validity check will do too!

samuelfangjw commented on src/main/java/teammates/ui/webapi/FeedbackSessionClosingRemindersAction.java:
```diff
@@ -32,6 +35,40 @@ public JsonResult execute() {
+                continue;
+            }
+
+            List<EmailWrapper> emailsToBeSent =
+                    emailGenerator.generateFeedbackSessionClosingWithExtensionEmails(feedbackSession, deadlineExtensions);
+            taskQueuer.scheduleEmailsForSending(emailsToBeSent);
```
Why should it be sent after? I don't believe there is any difference here.

moziliar replied:
Will the update always succeed? It is wrapped inside a `try-catch` block -> not really assuring to me.

moziliar replied:
Ok I see.. this is because `DeadlineExtension` is not the source of truth... My bad.

samuelfangjw commented on src/test/java/teammates/storage/api/DeadlineExtensionsDbTest.java:
```diff
@@ -0,0 +1,432 @@
+                .contains(DeadlineExtensionAttributes.valueOf(deadlineExtensionEmailSent)));
+
+        for (DeadlineExtensionAttributes deadlineExtension : deadlineExtensionsNeedingClosing) {
+            assertTrue(deadlineExtension.getEndTime().isAfter(Instant.now().minusSeconds(60)));
+            assertTrue(deadlineExtension.getEndTime().isBefore(TimeHelper.getInstantDaysOffsetFromNow(1).plusSeconds(60)));
```
60 has semantic meaning here (one minute). I don't believe there is much to gain by abstracting this out?

moziliar replied:
Ok, makes sense.

samuelfangjw commented on src/main/java/teammates/ui/webapi/FeedbackSessionClosingRemindersAction.java:
```diff
@@ -32,6 +35,40 @@ public JsonResult execute() {
+                            .build();
+                    logic.updateDeadlineExtension(updateOptions);
+                }
+            } catch (Exception e) {
```
Change to sub-types.

samuelfangjw commented on src/main/java/teammates/ui/webapi/DeadlineExtensionsWorkerAction.java:
```diff
@@ -0,0 +1,156 @@
+                logic.getDeadlineExtension(course.getId(), session.getFeedbackSessionName(), userEmail, isInstructor);
+        logic.deleteDeadlineExtension(course.getId(), session.getFeedbackSessionName(), userEmail, isInstructor);
+
+        if (deadlineExtension == null) {
+            log.severe("Unexpected error while revoking deadline extension");
+            return null;
+        }
```
Actually, nvm. I think we can move it up, since the delete will fail anyways. Will make the change.

moziliar reviewed (approved):
I believe the rest LGTM 👍
Thanks for the huge effort in getting the email working.

jayasting98 commented:
I will look through this again soon to double check the whole thing; I will need to find some period of time in the future to dedicate to this. Though I think when I first looked through, it was pretty much mostly good to go already.

I think the only thing I noted when I was first reviewing that may be worth discussing right now is the naming. I believe it was mentioned previously that the `DeadlineExtension` entity can also easily support early deadlines. In that case, would it be more apt to name it something more general like just `Deadline`? I don't think it may be super necessary though, so if you think the naming is fine as it is then that is probably alright I guess.

samuelfangjw commented:
> I will look through this again soon to double check the whole thing; I will need to find some period of time in the future to dedicate to this. Though I think when I first looked through, it was pretty much mostly good to go already.

No worries, this PR is not a blocker at the moment.

> I think the only thing I noted when I was first reviewing that may be worth discussing right now is the naming. I believe it was mentioned previously that the `DeadlineExtension` entity can also easily support early deadlines. In that case, would it be more apt to name it something more general like just `Deadline`? I don't think it may be super necessary though, so if you think the naming is fine as it is then that is probably alright I guess.

I don't really have issues with the naming, can change if it's better.

samuelfangjw commented:
I will update the PR tomorrow to remove "Create and Delete Deadline Extension API for use in future e2e tests". I realized that these are not needed for any e2e tests.

halfwhole commented on src/main/java/teammates/ui/webapi/DeadlineExtensionsWorkerAction.java:
```diff
@@ -0,0 +1,156 @@
+        revokeDeadlineExtensions(course, session,
+                deadlineExtensionRequest.getStudentExtensionsToRevoke(), false, notifyUsers);
+        revokeDeadlineExtensions(course, session,
+                deadlineExtensionRequest.getInstructorExtensionsToRevoke(), true, notifyUsers);
```
I'm not really familiar with task queues, but was wondering about the rationale for both handling deadline extension entities and sending emails under this same worker action - are there benefits to doing this, as compared to say directly creating/updating/deleting deadline extensions when we update a feedback session, and delegating only the part on sending emails to the task queue?

samuelfangjw replied:
Thanks for the review! Rationale here is that it might take awhile to update/create the entities, so if we delegate it to a task queue the user doesn't have to wait for all entities to be created/updated. We can do this because the DeadlineExtension entity is not the source of truth (the `studentDeadlines` and `instructorDeadlines` fields are already updated, and these will be used as the single source of truth).

halfwhole replied:
Makes sense, let's stick with this then!

halfwhole commented on src/main/java/teammates/ui/webapi/DeadlineExtensionsWorkerAction.java:
```diff
@@ -0,0 +1,156 @@
+                deadlineExtensionRequest.getStudentExtensionsToRevoke(), false, notifyUsers);
+        revokeDeadlineExtensions(course, session,
+                deadlineExtensionRequest.getInstructorExtensionsToRevoke(), true, notifyUsers);
+
```
Also I'm not sure about this, it's probably quite minor, but could we benefit from collating all the emails together and scheduling them in one shot? In terms of having a better spacing between emails

samuelfangjw replied:
I think we can! Actually, originally that was how I designed it but it was really messy. I've simplified it quite a bit since my first attempt, will try again to see if it's possible.

samuelfangjw replied:
Should be resolved, can see if it's ok.

halfwhole reviewed (commented):
Looks great, just have a few comments/questions!
I've only partially looked through the tests so far, so can't comment on that yet

samuelfangjw commented on src/main/java/teammates/logic/core/DeadlineExtensionsLogic.java:
```diff
@@ -0,0 +1,116 @@
+     */
+    public DeadlineExtensionAttributes updateDeadlineExtension(DeadlineExtensionAttributes.UpdateOptions updateOptions)
+            throws InvalidParametersException, EntityDoesNotExistException {
+        return deDb.updateDeadlineExtension(updateOptions);
```
Yes, thanks for spotting this!

samuelfangjw replied:
Added some logic to check if `endTime` update option is present. I update `sentClosingEmail` if it is not already explicit set. 

This is slightly different from the one in feedback session logic. I didn't like that one as it required an additional database read for information we can already get from the update options (with minor modifications).

samuelfangjw commented on src/main/java/teammates/common/util/TimeHelper.java:
```diff
@@ -39,6 +39,16 @@ public static Instant getInstantDaysOffsetBeforeNow(long offsetInDays) {
+    public static Instant getInstantHoursOffsetFromNow(long offsetInHours) {
+        return Instant.now().plus(Duration.ofHours(offsetInHours));
+    }
+
```
Wasn't aware of this class, will make the change. Originally, this method was used in production as well (for closed emails), but it's not needed now.

samuelfangjw commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -295,6 +297,7 @@ private Const() {
 
         public static final String DATABUNDLE = URI_PREFIX + "/databundle";
         public static final String DATABUNDLE_DOCUMENTS = URI_PREFIX + "/databundle/documents";
+        public static final String DEADLINE_EXTENSION = URI_PREFIX + "/deadline";
```
Removed this in latest commit.

samuelfangjw commented:
> I will update the PR tomorrow to remove "Create and Delete Deadline Extension API for use in future e2e tests". I realized that these are not needed for any e2e tests.

This has been removed! I don't think any of the changes are controversial, except maybe this part:
https://github.com/TEAMMATES/teammates/blob/44a940760c3abc17952cc7b658cbb85babb8d7b0/src/e2e/java/teammates/e2e/cases/BaseE2ETestCase.java#L317-L320

Any issues with throwing this particular exception for unimplemented method?

samuelfangjw commented:
> I think the only thing I noted when I was first reviewing that may be worth discussing right now is the naming. I believe it was mentioned previously that the `DeadlineExtension` entity can also easily support early deadlines. In that case, would it be more apt to name it something more general like just `Deadline`? I don't think it may be super necessary though, so if you think the naming is fine as it is then that is probably alright I guess.

@halfwhole, @moziliar any thoughts on this? Do you prefer `Deadline` instead of `DeadlineExtension`? I'm ok with changing  this if it's better (shouldn't take too long)

moziliar commented:
> Do you prefer Deadline instead of DeadlineExtension? I'm ok with changing this if it's better (shouldn't take too long)

@jayasting98 @samuelfangjw Good point raised!

Just thinking if "Deadline", or even "DeadlineExtension", is too generic. The thing I feel about `DeadlineExtensionEntity` is that we sometimes confuse it with the actual deadline information needed to verify if a submission (past deadline) is allowed. 

Am wondering if `SpecialDeadlineShadowEntity` works, since
1. the deadline is "special" (not the same as session closed time)
2. it is a deadline
3. it shadows the real information kept in `FeedbackSession`

moziliar commented:
> Any issues with throwing this particular exception for unimplemented method?

@samuelfangjw LGTM. This can hint to any user of this API that it is not meant for client side

samuelfangjw commented:
> @halfwhole did I miss something here? I feel like this entity should not be touched by client side

Yes, should not. But method still has to be overridden in `BaseE2ETestCase` (without major changes the base test case).

samuelfangjw commented:
> > Do you prefer Deadline instead of DeadlineExtension? I'm ok with changing this if it's better (shouldn't take too long)
> 
> @jayasting98 @samuelfangjw Good point raised!
> 
> Just thinking if "Deadline", or even "DeadlineExtension", is too generic. The thing I feel about `DeadlineExtensionEntity` is that we sometimes confuse it with the actual deadline information needed to verify if a submission (past deadline) is allowed.
> 
> Am wondering if `SpecialDeadlineShadowEntity` works, since
> 
> 1. the deadline is "special" (not the same as session closed time)
> 2. it is a deadline
> 3. it shadows the real information kept in `FeedbackSession`

Been thinking about this, I would actually prefer `DeadlineExtension` if this was the alternative. Special and shadow might add more ambiguity (what is special about this? what is shadow in this context?). I don't think there is a need at the moment to make it more specific, because it's unlikely to be confused with anything else.

moziliar commented:
I am fine with leaving the name as is for now. It is not a blocker to me.

halfwhole reviewed (approved):
LGTM!
As for the naming, I'm good with how it currently is as well - either "deadline" or "deadline extension" would sound alright, and email-wise, the use case should make it clear what it's for, and if there's documentation to back it up as well there should be no worries

samuelfangjw commented:
@moziliar, @halfwhole

Apologies for making changes after approval. I've kept this to one commit so it's easier to review : https://github.com/TEAMMATES/teammates/pull/11602/commits/2c27de05acd5a3d3b68e19ee3a225b4def0c1002.

Summary of changes:
1. Rename `DeadlineExtensionRequest` to `DeadlineExtensionsRequest`.
2. Changed body of DeadlineExtensionRequest to the following:
<img width="381" alt="Screen Shot 2022-03-23 at 7 08 25 PM" src="https://user-images.githubusercontent.com/60355570/159686168-9d227528-5624-46dd-b000-521ab8eef2e5.png">
Rationale (also see previous comment):

- No more need for `UpdateFeedbackSessionAction` to pre-process the deadline extensions into created/updated/deleted maps.
- We now pass the old deadlines into the worker, so we don't have to fetch this information again, simplifying implementation as well as reducing the number of db reads per deadline extension entity modified.
3. I've moved the logic to convert deadlines from long to instant (and instant to long) to within `DeadlineExtensionsRequest`. Feels cleaner to abstract it away so I don't have to deal with conversion in `DeadlineExtensionsWorkerAction`.
4. Added a test for `DeadlineExtensionsRequest`. I wrote that testcase to debug an issue, thought no harm leaving it there.

jayasting98 commented on src/main/java/teammates/storage/api/DeadlineExtensionsDb.java:
```diff
@@ -0,0 +1,227 @@
+            return false;
+        }
+
+        // Valid combination of parameters:
+        // courseId
+        // courseId, feedbackSessionName
+        // courseId, isInstructor, userEmail
+        return !query.isIsInstructorPresent() && !query.isUserEmailPresent() && !query.isFeedbackSessionNamePresent()
+                || !query.isIsInstructorPresent() && !query.isUserEmailPresent() && query.isFeedbackSessionNamePresent()
+                || query.isIsInstructorPresent() && query.isUserEmailPresent() && !query.isFeedbackSessionNamePresent();
```
Maybe we should use variables here instead of comments, to make it self-documenting?
```suggestion
        boolean isValidForDeletingAllInCourse = !query.isIsInstructorPresent() && !query.isUserEmailPresent() && !query.isFeedbackSessionNamePresent();
        boolean isValidForDeletingAllInSession = !query.isIsInstructorPresent() && !query.isUserEmailPresent() && query.isFeedbackSessionNamePresent();
        boolean isValidForDeletingAllForUser = query.isIsInstructorPresent() && query.isUserEmailPresent() && !query.isFeedbackSessionNamePresent();
        return isValidForDeletingAllInCourse || isValidForDeletingAllInSession || isValidForDeletingAllForUser;
```

samuelfangjw replied:
Sure, I'll update now.

jayasting98 reviewed (commented):
Just one last nit possibly.

jayasting98 reviewed (approved):
Looks good!

moziliar commented on src/main/java/teammates/ui/webapi/DeadlineExtensionsWorkerAction.java:
```diff
@@ -28,129 +28,120 @@
+                .map(entry ->
+                        emailGenerator.generateDeadlineExtensionEmail(course, session, oldDeadlines.get(entry.getKey()),
+                                entry.getValue(), EmailType.DEADLINE_EXTENSION_UPDATED, entry.getKey(), areInstructors))
+                .collect(Collectors.toList());
```
Tiny nit: would it be better to have `getDeadlineOpsEmails` that takes in the `EmailType` as input so as to reduce code duplication?

samuelfangjw replied:
I can try, but the difference is not as simple as different EmailType. Notice the start and end times are different for each method. Also, getDeadlineUpdatedEmails takes in an extra parameter.

moziliar replied:
Nvm! Can leave it!

moziliar reviewed (changes requested):
LGTM! One final nit for your consideration.

moziliar reviewed (approved):
LGTM!

moziliar commented:
@halfwhole would you be free to take a look at the latest commit before merging?
